libraries/shared/watcher/storage_watcher.go

```diff
@@ -17,73 +17,103 @@
 package watcher
 
 import (
+	"fmt"
 	"reflect"
-	"strings"
+	"time"
 
 	"github.com/ethereum/go-ethereum/common"
 	"github.com/sirupsen/logrus"
 
+	"github.com/vulcanize/vulcanizedb/libraries/shared/fetcher"
 	"github.com/vulcanize/vulcanizedb/libraries/shared/storage"
 	"github.com/vulcanize/vulcanizedb/libraries/shared/storage/utils"
 	"github.com/vulcanize/vulcanizedb/libraries/shared/transformer"
 	"github.com/vulcanize/vulcanizedb/pkg/datastore/postgres"
-	"github.com/vulcanize/vulcanizedb/pkg/fs"
 )
 
 type StorageWatcher struct {
-	db           *postgres.DB
-	tailer       fs.Tailer
-	Queue        storage.IStorageQueue
-	Transformers map[common.Address]transformer.StorageTransformer
+	db             *postgres.DB
+	StorageFetcher fetcher.IStorageFetcher
+	Queue          storage.IStorageQueue
+	Transformers   map[common.Address]transformer.StorageTransformer
 }
 
-func NewStorageWatcher(tailer fs.Tailer, db *postgres.DB) StorageWatcher {
+func NewStorageWatcher(fetcher fetcher.IStorageFetcher, db *postgres.DB) StorageWatcher {
 	transformers := make(map[common.Address]transformer.StorageTransformer)
 	queue := storage.NewStorageQueue(db)
 	return StorageWatcher{
-		db:           db,
-		tailer:       tailer,
-		Queue:        queue,
-		Transformers: transformers,
+		db:             db,
+		StorageFetcher: fetcher,
+		Queue:          queue,
+		Transformers:   transformers,
 	}
 }
 
-func (watcher StorageWatcher) AddTransformers(initializers []transformer.StorageTransformerInitializer) {
+func (storageWatcher StorageWatcher) AddTransformers(initializers []transformer.StorageTransformerInitializer) {
 	for _, initializer := range initializers {
-		storageTransformer := initializer(watcher.db)
-		watcher.Transformers[storageTransformer.ContractAddress()] = storageTransformer
+		storageTransformer := initializer(storageWatcher.db)
+		storageWatcher.Transformers[storageTransformer.ContractAddress()] = storageTransformer
 	}
 }
 
-func (watcher StorageWatcher) Execute() error {
-	t, tailErr := watcher.tailer.Tail()
-	if tailErr != nil {
-		return tailErr
+func (storageWatcher StorageWatcher) Execute(rows chan utils.StorageDiffRow, errs chan error, queueRecheckInterval time.Duration) {
+	ticker := time.NewTicker(queueRecheckInterval)
+	go storageWatcher.StorageFetcher.FetchStorageDiffs(rows, errs)
+	for {
+		select {
+		case fetchErr := <-errs:
+			logrus.Warn(fmt.Sprintf("error fetching storage diffs: %s", fetchErr))
+		case row := <-rows:
+			storageWatcher.processRow(row)
+		case <-ticker.C:
+			storageWatcher.processQueue()
+		}
```

Comment:
Would it make sense to add a default sleep case here if all falls through?

Reply:
🤔 good question! Curious to hear more from you and others.

On the one hand, I'm inclined to keep spinning so that we stay up to date with new rows as quickly as possible. But, on the other hand, I could see the benefit of throttling that if we're checking many times more often than necessary.

I think I'd be tempted to stick with the current implementation and then do some investigation around optimal sleep times if we decide to go that route, but am very open to reconsidering.

Reply:
I'm thinking we run this on the same machine as the other commands, won't this hog a core basically permanently?

Reply:
Like when the interval is 5 minutes, this will still loop at blazing speed until some channel is populated no?

Reply:
that sounds right but it's unclear to me if we need to make that optimization right now. Our [light sync code](https://github.com/vulcanize/vulcanizedb/blob/staging/cmd/lightSync.go#L84), for example, has a similar pattern, and seems to run fine alongside other processes right now (granted, that loop does have a sleep in one of the `case` statements - but I believe it would loop by default while neither the ticker or nor the missing blocks are receiving messages).

I'm definitely not opposed to exploring what kind of hit we're taking here and putting a story on the board to add defaults that sleep. But i think it may be worth a separate story because it's a bit tricky to test.

Reply:
That linked code is a bit different, since it'll sleep after _every_ run of backfilling if it yielded no missing blocks. Only in the case of actual meat to work with it'll re-run directly afterwards. Anyway, putting it in a later story is fine :)